File: dart/lib/scheme/scheme_neutral.dart
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import 'package:material_color_utilities/hct/hct.dart';
import 'package:material_color_utilities/palettes/tonal_palette.dart';
import 'dynamic_scheme.dart';
import 'variant.dart';

/// A Dynamic Color theme that is near grayscale.
class SchemeNeutral extends DynamicScheme {
  SchemeNeutral({
    required Hct sourceColorHct,
    required super.isDark,
    required super.contrastLevel,
  }) : super(
          sourceColorArgb: sourceColorHct.toInt(),
          variant: Variant.neutral,
          primaryPalette: TonalPalette.of(sourceColorHct.hue, 12.0),
          secondaryPalette: TonalPalette.of(sourceColorHct.hue, 8.0),
          tertiaryPalette: TonalPalette.of(sourceColorHct.hue, 16.0),
          neutralPalette: TonalPalette.of(sourceColorHct.hue, 2.0),
          neutralVariantPalette: TonalPalette.of(sourceColorHct.hue, 2.0),
        );
}

File: cpp/scheme/scheme_neutral.cc
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme_neutral.h"

#include "cpp/cam/hct.h"
#include "cpp/dynamiccolor/dynamic_scheme.h"
#include "cpp/dynamiccolor/variant.h"
#include "cpp/palettes/tones.h"

namespace material_color_utilities {

SchemeNeutral::SchemeNeutral(Hct set_source_color_hct, bool set_is_dark,
                             double set_contrast_level)
    : DynamicScheme(
          /*source_color_argb:*/ set_source_color_hct.ToInt(),
          /*variant:*/ Variant::kNeutral,
          /*contrast_level:*/ set_contrast_level,
          /*is_dark:*/ set_is_dark,
          /*primary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 12.0),
          /*secondary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 8.0),
          /*tertiary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 16.0),
          /*neutral_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 2.0),
          /*neutral_variant_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 2.0)) {}

SchemeNeutral::SchemeNeutral(Hct set_source_color_hct, bool set_is_dark)
    : SchemeNeutral::SchemeNeutral(set_source_color_hct, set_is_dark, 0.0) {}

}  // namespace material_color_utilities
